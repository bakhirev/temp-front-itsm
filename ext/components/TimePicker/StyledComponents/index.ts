export * from './IconWrapper';
export * from './Wrapper';
