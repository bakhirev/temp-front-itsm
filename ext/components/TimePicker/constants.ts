export const INPUT_HEIGHT = {
  micro: 55,
  small: 63,
  big: 80,
};
export const INPUT_WIDTH = 200;
export const MENU_HEIGHT = 208;
export const MENU_WIDTH = 128;
export const MENU_LEFT_RIGHT_BORDER_PADDING = 16;
export const MENU_TOP_BOTTOM_BORDER_PADDING = 8;

export const DROPDOWN_OFFSET = INPUT_HEIGHT;
