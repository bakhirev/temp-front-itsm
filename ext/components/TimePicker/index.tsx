import React, { useState, FC, useRef, FocusEvent, useCallback } from 'react';

import type { Size } from '../Inputs/common/types';
import { getIconSize } from '../Inputs/common/utils';
import { InputMask } from '../Inputs/Mask';
import { ReactComponent as TimeOutline } from '../Icons/system/TimeOutline.svg';

import { ADDITIONAL_MESSAGE, DEFAULT_WIDTH, MASK } from './components/constants';
import { IconWrapper, Wrapper } from './StyledComponents';
import {
  timeValid,
  parseValue,
  formatValue,
  validateUserTimeRange,
  activeTimeParse,
} from './utils';
import { DropdownMenu } from './components/DropdownMenu';

export type TTime = {
  hours: string;
  minutes: string;
};

export interface ITimePickerProps {
  /** Значение компонента */
  value: TTime;
  /** Коллбек на изменение состояния */
  onChange: (value: TTime) => void;
  /** Лэйбл к инпуту */
  label?: string;
  /** Размер инпута */
  size?: Size;
  /** Флаг блокировки */
  disabled?: boolean;
  /** Коллбэк на событие фокуса инпута */
  onFocus?: (event: FocusEvent<HTMLInputElement>) => void;
  /** Коллбэк на событие потери фокуса инпута */
  onBlur?: (event: FocusEvent<HTMLInputElement>) => void;
  /** Минимальное значение интервала селекта */
  selectRangeStart?: TTime;
  /** Максимальное значение интервала селекта */
  selectRangeEnd?: TTime;
  /** data-test-id атрибут для тестирования компонента */
  dataTestId?: string;
  /** Имя класса для переопределения стилей */
  className?: string;
  /** Уникальный идентификатор */
  id?: string;
  /** Ширина инпута */
  width?: string | number;
  /** Дополнительный текст под инпутом */
  additionalText?: string;
}

export const TimePicker: FC<ITimePickerProps> = ({
  value = { hours: '', minutes: '' },
  onChange,
  selectRangeStart,
  selectRangeEnd,
  dataTestId,
  className,
  size = 'big',
  disabled,
  additionalText = ADDITIONAL_MESSAGE,
  onBlur,
  onFocus,
  width = DEFAULT_WIDTH,
  id,
  ...props
}) => {
  const inputRef = useRef<any>(null);
  const [error, setError] = useState(false);
  const [menuOpen, setMenuOpen] = useState<boolean>(false);

  const valueInput = parseValue(value);

  const handleChange = (value: { value: string; masked: string }) => {
    setMenuOpen(false);
    onChange && onChange(formatValue(value));
  };

  const handleFocus = (event: FocusEvent<HTMLInputElement>) => {
    onFocus && onFocus(event);
    setError(false);
  };

  const handleBlur = (event: FocusEvent<HTMLInputElement>) => {
    setError(timeValid(value, valueInput));
    onBlur && onBlur(event);
    setMenuOpen(false);
  };

  const handleComplete = useCallback((value: string) => {
    onTimeSelect(value);
  }, []);

  const onTimeSelect = (time: string) => {
    const value = time.replace(/\D+/g, '');

    onChange &&
      onChange({
        hours: `${value[0]}${value[1]}`,
        minutes: `${value[2]}${value[3]}`,
      });
    setMenuOpen(false);
  };

  const handleClickIcon = () => {
    if (menuOpen) {
      setMenuOpen(false);
    } else {
      setMenuOpen(true);
      inputRef.current.focus();
    }
  };

  const { rangeStart, rangeEnd } = validateUserTimeRange(selectRangeStart, selectRangeEnd);
  const iconSize = getIconSize(size);

  const icon = (
    <IconWrapper disabled={disabled} active={menuOpen} onMouseDown={(e) => e.preventDefault()}>
      <TimeOutline
        width={iconSize}
        height={iconSize}
        onClick={disabled ? undefined : handleClickIcon}
      />
    </IconWrapper>
  );
  return (
    <Wrapper dataTestId={dataTestId} className={className} width={width}>
      <InputMask
        icon={icon}
        id={id}
        mask={MASK}
        onFocus={handleFocus}
        disabled={disabled}
        refInput={inputRef}
        onBlur={handleBlur}
        status={error ? 'error' : 'default'}
        additionalText={error ? additionalText : ''}
        onComplete={handleComplete}
        size={size}
        value={valueInput}
        onChange={handleChange}
        alwaysShowMask
        width={width}
        {...props}
      />
      <DropdownMenu
        size={size}
        open={menuOpen}
        onTimeSelect={onTimeSelect}
        activeTime={activeTimeParse(value)}
        rangeStart={rangeStart}
        rangeEnd={rangeEnd}
      />
    </Wrapper>
  );
};
