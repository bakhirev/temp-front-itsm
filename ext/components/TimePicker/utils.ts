import { DEFAULT_SELECT_RANGE_END, DEFAULT_SELECT_RANGE_START } from './components/constants';

import type { TTime } from './index';

const HOURS_ITEMS = Array.from(Array(24).keys())
  .map(String)
  .map((item) => (item.length === 1 ? `0${item}` : item));

export const parseRangeTime = (time?: TTime) => {
  if (!(time && time.hours && HOURS_ITEMS.includes(time.hours))) return null;

  const { hours, minutes } = time;
  return { hours, minutes: minutes === '00' || minutes === '30' ? minutes : '00' };
};

export const validateUserTimeRange = (rawRangeStart?: TTime, rawRangeEnd?: TTime) => {
  const rangeStart = parseRangeTime(rawRangeStart) || DEFAULT_SELECT_RANGE_START;
  const rangeEnd = parseRangeTime(rawRangeEnd) || DEFAULT_SELECT_RANGE_END;

  if (rangeStart.hours > rangeEnd.hours) {
    return {
      rangeStart,
      rangeEnd: DEFAULT_SELECT_RANGE_END,
    };
  } else if (rangeStart.hours === rangeEnd.hours && rangeStart.minutes > rangeEnd.minutes) {
    return { rangeStart, rangeEnd: rangeStart };
  }

  return { rangeStart, rangeEnd };
};

export const activeTimeParse = (value: { hours: string; minutes: string }) => {
  return value.hours + ' : ' + value.minutes;
};

export const timeValid = (value: { hours: string; minutes: string }, valueInput: string) => {
  if (+value.hours > 24 || +value.minutes > 59 || (valueInput.length < 4 && valueInput !== '')) {
    return true;
  } else {
    return false;
  }
};

export const parseValue = (value: { hours: string; minutes: string }) => {
  const newValue = value.hours + value.minutes;

  return newValue;
};

export const formatValue = (str: { value: string; masked: string }) => {
  return {
    hours: str.value?.slice(0, 2),
    minutes: str.value?.slice(2, 4),
  };
};
