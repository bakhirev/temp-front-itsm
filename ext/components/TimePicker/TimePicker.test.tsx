import React from 'react';
import { shallow } from 'enzyme';

import { TimePicker } from '../TimePicker';

jest.mock('../common', () => ({
  DEFAULT_THEME: {},
}));

const textError = `Ошибка`;
const label = 'Выберите время';
const changeUsernameSpy = jest.fn();

describe('TimePicker', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render component with props', () => {
    const wrapper = shallow(
      <TimePicker
        value={{ hours: '', minutes: '' }}
        onChange={(value: any) => value}
        label={label}
        additionalText={textError}
        size={'big'}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should call onChange when user presses keys in timepicker field', () => {
    const wrapper = shallow(
      <TimePicker
        value={{ hours: '', minutes: '' }}
        onChange={changeUsernameSpy}
        label={label}
        size={'small'}
      />
    );
    const input = wrapper.childAt(0);
    input.simulate('change', { value: '1100', masked: '' });

    expect(changeUsernameSpy).toBeCalledTimes(1);
    expect(changeUsernameSpy).toBeCalledWith({ hours: '11', minutes: '00' });
  });

  it('should call onChange when user presses keys in input timepicker', () => {
    const wrapper = shallow(
      <TimePicker
        value={{ hours: '', minutes: '' }}
        onChange={changeUsernameSpy}
        label={label}
        size={'small'}
      />
    );
    const input = wrapper.childAt(0);

    input.simulate('change', { value: '009992313', masked: '' });

    expect(changeUsernameSpy).toBeCalledTimes(1);
    expect(changeUsernameSpy).toBeCalledWith({ hours: '00', minutes: '99' });
  });
});
