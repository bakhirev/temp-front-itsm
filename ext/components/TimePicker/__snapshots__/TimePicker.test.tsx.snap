// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimePicker should render component with props 1`] = `
<styled.div
  width="100%"
>
  <Component
    additionalText=""
    alwaysShowMask={true}
    icon={
      <ForwardRef(styled.div)
        active={false}
        onMouseDown={[Function]}
        theme={Object {}}
      >
        <div
          height={24}
          onClick={[Function]}
          width={24}
        />
      </ForwardRef(styled.div)>
    }
    label="Выберите время"
    mask="99 : 99"
    onBlur={[Function]}
    onChange={[Function]}
    onComplete={[Function]}
    onFocus={[Function]}
    refInput={
      Object {
        "current": null,
      }
    }
    size="big"
    status="default"
    value=""
    width="100%"
  />
  <Component
    activeTime=" : "
    onTimeSelect={[Function]}
    open={false}
    rangeEnd={
      Object {
        "hours": "23",
        "minutes": "30",
      }
    }
    rangeStart={
      Object {
        "hours": "00",
        "minutes": "00",
      }
    }
    size="big"
  />
</styled.div>
`;
